E2e tests rolling updates

gflarity opened the pull request:
PR for Rolling Updates tests. I was able to get most of them working

1. Please take a look at the fix in rollingupdate.go. Seems like a simple enough typo/cut and pass issue?
2. These fails are failing still:
    * 10 (always fails) and
    * 18 fails intermittently

shayasoolin commented on operator/e2e/tests/setup.go:
```diff
@@ -324,12 +340,16 @@ func waitForPodConditions(tc TestContext, expectedTotalPods, expectedPending int
-func scalePCSGAndWait(tc TestContext, pcsgName string, replicas int32, expectedTotalPods, expectedPending int) {
+// scalePCSGAndWait attempts to scale a PCSG via the PCS template and waits for the expected pod conditions.
+// WARNING: This only affects NEW PCS replicas created after the change. Existing PCSG instances are NOT updated.
+// DEPRECATED: Use scalePCSGAcrossAllReplicasAndWait instead to scale existing PCSG instances directly.
```
Why keep this function here if it's deprecated and unused?

shayasoolin commented on operator/e2e/utils/k8s_client.go:
```diff
@@ -485,6 +485,75 @@ func ScalePodCliqueScalingGroupWithClient(ctx context.Context, dynamicClient dyn
 }
 
+// ScalePCSGInPCSWithClient modifies the PCS template's podCliqueScalingGroups[].replicas field.
+// WARNING: This only affects NEW PCS replicas created after the change. Existing PCSG instances
```
Is this function in use currently? If not, can we remove it?

gflarity replied:
Yes, thanks. It didn't show up in linter because it's capitalized.

shayasoolin commented on operator/e2e/tests/rolling_updates_test.go:
```diff
@@ -0,0 +1,2670 @@
+	logger.Info("🎉 Rolling Update with PodClique scale-in before update test (RU-21) completed successfully!")
+}
+
+// podEvent represents a pod lifecycle event during rolling update
```
Test-infra code from here till the end of this file better be split to a utils files.

shayasoolin commented on operator/e2e/tests/rolling_updates_test.go:
```diff
@@ -0,0 +1,2670 @@
+// verifyOnePodDeletedAtATime verifies only one pod globally is being deleted at a time
+// during rolling updates.
+//
+// This function handles the fact that Kubernetes watch events can arrive out of order.
```
I'm curious if KAI-scheduler has similar tests where it's required to check the order of events, and how these validations are implemented there. @romanbaron

gflarity commented on operator/e2e/tests/rolling_updates_test.go:
```diff
@@ -0,0 +1,2670 @@
+//  1. The old pod is deleted (e.g., hostname: "workload1-pc-a-0-0")
+//  2. The PodClique remains unchanged (same name and replica count)
+//  3. The new pod is created to fill the same logical position
+//  4. The new pod receives the same hostname (e.g., "workload1-pc-a-0-0")
```
AH, this isn't the Node host name. This pod.Spec.Hostname, he internal hostname of the pod itself. it's what the hostname command would return inside the container. See /internal/controller/podclique/components/pod/pod.go:

```go
// configurePodHostname sets the pod hostname and subdomain for service discovery
func configurePodHostname(pcsName string, pcsReplicaIndex int, pclqName string, pod *corev1.Pod, podIndex int) {
	// Set hostname for service discovery (e.g., "my-pclq-0")
	pod.Spec.Hostname = fmt.Sprintf("%s-%d", pclqName, podIndex)

	// Set subdomain to headless service name (reusing existing logic)
	pod.Spec.Subdomain = apicommon.GenerateHeadlessServiceName(
		apicommon.ResourceNameReplica{Name: pcsName, Replica: pcsReplicaIndex})
}
```
But I definitely agree this is confusing. I'll update the comment to give a better explanation.

gflarity commented:
Thanks @shayasoolin, please take a look: 

* removed the unused functions I missed
* moved the helper functions into separate files 
* reduced repetition in the tests themselves

T

shmuel-runai commented on operator/e2e/tests/debug_utils.go:
```diff
@@ -0,0 +1,170 @@
+}
+
+// splitLines splits a string into lines
+func splitLines(s string) []string {
```
what is wrong with `strings.Split(s, "\n")`

shmuel-runai commented on operator/e2e/tests/debug_utils.go:
```diff
@@ -0,0 +1,170 @@
+						for _, line := range splitLines(lines) {
+							if len(line) > 0 {
+								// Print all lines, but highlight rolling update related ones
+								if containsRollingUpdateTag(line) {
```
a bit YANGI but,`containsRollingUpdateTag` should be a function param to this function (filterFN), so that this captureOperatorLogs could be reused with other testing than rolling update

shmuel-runai commented on operator/e2e/tests/rolling_update_utils.go:
```diff
@@ -0,0 +1,1149 @@
+// patchPCSWithSIGTERMIgnoringCommand patches all containers in the PCS to use a command that ignores SIGTERM
+// and sets the termination grace period to 5 seconds. This makes pods ignore graceful shutdown but still
+// allows rolling updates to progress in a reasonable time for testing.
+// Uses tc.Workload.Name as the PCS name.
```
nit: last line of the comment is redundant

shmuel-runai commented on operator/internal/controller/podclique/components/pod/rollingupdate.go:
```diff
@@ -60,7 +60,7 @@ func (w *updateWork) getPodNamesPendingUpdate(deletionExpectedPodUIDs []types.UI
 func (w *updateWork) getNextPodToUpdate() *corev1.Pod {
 	if len(w.oldTemplateHashReadyPods) > 0 {
-		slices.SortFunc(w.oldTemplateHashPendingPods, func(a, b *corev1.Pod) int {
+		slices.SortFunc(w.oldTemplateHashReadyPods, func(a, b *corev1.Pod) int {
```
great catch

gflarity commented on operator/e2e/tests/rolling_update_utils.go:
```diff
@@ -0,0 +1,1149 @@
+// that completes when the rolling update finishes.
+// Uses tc.Workload.Name as the PCS name and tc.Timeout for the wait timeout.
+func triggerRollingUpdate(tc TestContext, expectedReplicas int32, cliqueNames ...string) *errgroup.Group {
+	g := new(errgroup.Group)
```
This needs to run asynchronously, see the call sites. This just seemed a like a clean way to do that without re-inventing the wheel.  However after a bit more research is seems like <-chan error is a bit more idiomatic. Either works for me.

shmuel-runai commented on operator/e2e/tests/debug_utils.go:
```diff
@@ -0,0 +1,156 @@
+				}
+
+				// Read logs
+				buf := make([]byte, 64*1024) // 64KB buffer
```
nit: should be a const, not a comment

shmuel-runai reviewed (approved):
Nice work
